### PR TITLE
fix: Information Popover is not keyboard accessible

### LINKED ISSUE #11714: The popover for old sub list still remains in the screen after switching to new sub section in B2B table (while keyboarding?)
**Tell us the version of Spartacus**
- Library version: 3.2.0-next.2, develop branch

**To Reproduce**

1. open unit list and click one row to enter unit detail page:

image

2. the (i) icon as the first focusable element in unit detail page is automatically focused. Press enter key, and you can see popover.

image

3.  click Users section to enter unit user sub section. Press enter key. See expected popover in user card:

image

4. Click Approver section, now the approver sub section is displayed. HOWEVER, the popover for Users is still displayed in the screen:

image

When we click the (i) icon in Approvers card, the corresponding popover can be displayed. Unfortunately, the popover for Users is still in the screen:

image

**Expected behavior**

When we switch from Users sub section to Approver sub section, the popover for former card should disappear.

**Desktop (please complete the following information):**
 - OS: windows 10
 - Browser: Chrome
 - Version 89.0.4389.90 (Official Build) (64-bit)

**Additional context**

### LINKED ISSUE #11774: Information Popover is not keyboard accessible (My COmpany and generally)
**Issues captured on video:**
https://video.sap.com/media/t/1_o07yiw6z
_See minutes 0:13 and 4:00_

- ESC key needs to be pressed 2x (twice) otherwise it won’t close, this is not an accessible pattern.
- If popover is opened with SPACEBAR, pressing ESC will trigger the skippers and it won’t close the popover.
- There is no other additional mechanism to close the popover other than the ESC key (Android gesture doesn’t have ESC), a closing button should be included.
- When opening popover with ENTER, popover should receive focus, next TAB should go to closing button. (mandatory)
- When opening popover with SPACEBAR, popover may not be focused but 1x ESC or Tabbing away should close the popover. (optional, if too much work stick to mandatory above)
- **We already have an accessible “details” popover for My Company, that can be used as a reference.**

Screen Shot 2021-04-01 at 9 15 09 AM
Screen Shot 2021-04-01 at 9 15 48 AM

### PR DESCRIPTION
**Pull request by Pucek9**
Closes #11774
Closes #11714

### CODE REVIEW

**Review by developpeurweb — CHANGES_REQUESTED** [review]
This seems to work fine, except when we use the `Spacebar`. So let's simplify, and make `Spacebar` do the same as the `Enter` key, please.
(I requested this differently for the details popover making "Read mode" considerations for future Screen Reader implementation, but I realize it gets us into collateral issues when trying to standardize, so let's simplify it.
Please ping me if any doubts.

**Comment by giancorderoortiz** [comment]
What is status of this pr? Eager to test fix.

**Comment by Pucek9** [comment]
> What is status of this pr? Eager to test fix.

it has been on hold due to other priorities. Now I can continue work on it

File: projects/storefrontlib/src/shared/components/popover/popover.directive.ts [hunk 1]
```diff
@@ -72,26 +73,57 @@ export class PopoverDirective {
...
+    if (event?.target === this.element.nativeElement && !this.isOpen) {
+      this.eventSubject.next(PopoverEvent.OPEN_BY_KEYBOARD);
+    } else if (this.isOpen) {
+      this.eventSubject.next(PopoverEvent.CLOSE_BUTTON_KEYDOWN);
+    }
   }
 
+  @HostListener('keydown.tab')
+  @HostListener('keydown.shift.tab')
+  handleTab(): void {
```
Comment (Matejk00) [inline projects/storefrontlib/src/shared/components/popover/popover.directive.ts:92]: Wouldn't be easier/cleaner to listen all keydown events and then recognise specific key inside?

**Review by Matejk00 — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/shared/components/popover/popover.directive.ts (hunk 1 above)
Comment (Pucek9) [reply projects/storefrontlib/src/shared/components/popover/popover.directive.ts:92]: as we agreed in discuss, I leave it like it was

File: projects/schematics/src/migrations/4_0/methods-and-properties-deprecations/data/popover-component.migration.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import {
+  INSIDE_CLICKED,
+  POPOVER_COMPONENT,
+  SPARTACUS_STOREFRONTLIB,
+  TODO_SPARTACUS,
+} from '../../../../shared/constants';
+import { MethodPropertyDeprecation } from '../../../../shared/utils/file-utils';
+
```
Comment (Matejk00) [inline projects/schematics/src/migrations/4_0/methods-and-properties-deprecations/data/popover-component.migration.ts:8]: Comment with path to file is missing.

File: projects/schematics/src/migrations/4_0/methods-and-properties-deprecations/data/popover-directive.migration.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import {
+  HANDLE_OPEN,
+  POPOVER_DIRECTIVE,
+  SPARTACUS_STOREFRONTLIB,
+  TODO_SPARTACUS,
+  TOGGLE,
+} from '../../../../shared/constants';
+import { MethodPropertyDeprecation } from '../../../../shared/utils/file-utils';
+
```
Comment (Matejk00) [inline projects/schematics/src/migrations/4_0/methods-and-properties-deprecations/data/popover-directive.migration.ts:9]: Same here. We can add a comment with path to file to be consistent with other files.

File: docs/migration/4_0.md [hunk 4]
```diff
@@ -1617,3 +1625,20 @@ Removed public members of `CxApi`: `CheckoutService`, `CheckoutDeliveryService`,
 ### UserIdService
 
 - `invokeWithUserId` method was removed. Use `takeUserId` instead.
+
+### PopoverDirective
+
+- Removed `PositioningService` from constructor.
+- PopoverDirective now implements `ngOnInit`
```
Comment (Matejk00) [inline docs/migration/4_0.md]: Missing backticks for PopoverDirective.

File: projects/storefrontlib/src/shared/components/popover/popover.directive.ts [hunk 5]
```diff
@@ -72,26 +72,56 @@ export class PopoverDirective {
...
+    if (!this.focusConfig?.trap && this.isOpen) {
+      this.eventSubject.next(PopoverEvent.CLOSE_BUTTON_KEYDOWN);
+    }
+  }
   /**
```
Comment (Matejk00) [inline projects/storefrontlib/src/shared/components/popover/popover.directive.ts]: We missing empty line here.

**Review by Matejk00 — CHANGES_REQUESTED** [review]
Can you take a look also on line 66 in `popover.directive.ts`? I think we need to remove such comment as is not referring to any method.

**Review by Matejk00 — APPROVED** [review]

**Review by developpeurweb — APPROVED** [review]

**Review by marlass — APPROVED** [review]

**Pull request merged**